### PR TITLE
tracing: Fixes issue with small LightStep reports.

### PR DESCRIPTION
**Pull request by rnburn**
*Description*:
Changes to use a larger default size for LightStep reports sent to satellites. Fixes https://github.com/lightstep/lightstep-tracer-cpp/issues/106

*Risk Level*: Low

### CODE REVIEW

**Comment by rnburn** [comment]
cc @jmacd

**Comment by ccaraman** [comment]
@rnburn Please fix tests and ping once ready for a review.

**Comment by rnburn** [comment]
@ccaraman - should be ready for review now.

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.cc [hunk 1]
```diff
@@ -150,8 +159,11 @@ LightStepDriver::LightStepDriver(const Json::Object& config,
     tls_options.use_thread = false;
     tls_options.use_single_key_propagation = true;
     tls_options.logger_sink = LightStepLogger{};
-    tls_options.max_buffered_spans = std::function<size_t()>{
-        [this] { return runtime_.snapshot().getInteger("tracing.lightstep.min_flush_spans", 5U); }};
+
+    tls_options.max_buffered_spans = std::function<size_t()>{[this] {
+      return runtime_.snapshot().getInteger("tracing.lightstep.min_flush_spans",
```
Comment (ccaraman) [inline source/extensions/tracers/lightstep/lightstep_tracer_impl.cc]: Will there be any issues experienced by users when switching to this higher value?

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.cc [hunk 2]
```diff
@@ -36,6 +37,14 @@ void LightStepLogger::operator()(lightstep::LogLevel level,
 LightStepDriver::LightStepTransporter::LightStepTransporter(LightStepDriver& driver)
     : driver_(driver) {}
 
+// If the default min_flush_spans value is too small, spans can be dropped between reports.
+// Hence, we choose a number that's large enough, but divide by the number of CPUs since Envoy
+// creates separate tracers for each thread.
+//
+// See https://github.com/lightstep/lightstep-tracer-cpp/issues/106
+const size_t LightStepDriver::default_min_flush_spans =
```
Comment (ccaraman) [inline source/extensions/tracers/lightstep/lightstep_tracer_impl.cc]: Should the concurrency value passed in through the command line be used first then default to the hardware concurrency if not specified?
https://github.com/envoyproxy/envoy/blob/c92a3017017dfa31241dec13dc6a1479090318d0/source/server/options_impl.cc#L53

Comment (jmacd) [reply source/extensions/tracers/lightstep/lightstep_tracer_impl.cc]: @rnburn I gave this more thought and now believe we should just fill in a constant here. Envoy's architecture is that every CPU is an independent worker, so as long as the amount of memory available scales with the number of processors, there's no reason to limit the buffering of LightStep spans according to processor count.

Will you adjust this to use a fixed constant of 200? That avoids the question Constance asks here.

Comment (rnburn) [reply source/extensions/tracers/lightstep/lightstep_tracer_impl.cc]: Update to use 200

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.h [hunk 3]
```diff
@@ -60,6 +60,8 @@ class LightStepDriver : public Common::Ot::OpenTracingDriver {
   Runtime::Loader& runtime() { return runtime_; }
   LightstepTracerStats& tracerStats() { return tracer_stats_; }
 
+  static const size_t default_min_flush_spans;
```
Comment (ccaraman) [inline source/extensions/tracers/lightstep/lightstep_tracer_impl.h]: nit: static const vars are all upper case

**Comment by ccaraman** [comment]
@htuch final pass please - LGTM - just one nit

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.h [hunk 4]
```diff
@@ -60,6 +60,8 @@ class LightStepDriver : public Common::Ot::OpenTracingDriver {
   Runtime::Loader& runtime() { return runtime_; }
   LightstepTracerStats& tracerStats() { return tracer_stats_; }
 
+  static const size_t DEFAULT_MIN_FLUSH_SPANS;
```
Comment (htuch) [inline source/extensions/tracers/lightstep/lightstep_tracer_impl.h]: Nit: Prefer `DefaultMinFlushSpans` for Envoy recommended constant style.

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.cc [hunk 5]
```diff
@@ -36,6 +36,13 @@ void LightStepLogger::operator()(lightstep::LogLevel level,
 LightStepDriver::LightStepTransporter::LightStepTransporter(LightStepDriver& driver)
     : driver_(driver) {}
 
+// If the default min_flush_spans value is too small, the larger number of reports can overwhelm
+// LightStep's sattelites. Hence, we need to choose a number that's large enough; though, it's
```
Comment (htuch) [inline source/extensions/tracers/lightstep/lightstep_tracer_impl.cc]: Nit: typo `sattelites`

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.h (hunk 4 above)
Comment (rnburn) [reply source/extensions/tracers/lightstep/lightstep_tracer_impl.h]: Renamed.

File: source/extensions/tracers/lightstep/lightstep_tracer_impl.cc (hunk 5 above)
Comment (rnburn) [reply source/extensions/tracers/lightstep/lightstep_tracer_impl.cc]: Fixed.

**Pull request merged**